### PR TITLE
Improve performance of Iceberg snapshot expiration

### PR DESCRIPTION
**Pull request by electrum**
## Description

Improve performance of the `expire_snapshots` operation when there are many snapshots that reference the same manifest files. Previously, the manifest files were read once for each snapshot.

We now use the Iceberg library code to compute the files to delete. The previous code may have correctness implications for cherry-picked commits or other scenarios.

## Documentation

(x) No documentation is needed.

## Release notes

(x) Release notes entries required with the following suggested text:

```markdown
# Iceberg connector
* Improve performance of the `expire_snapshots` command for tables with many snapshots. ({issue}`13399`)
```

### CODE REVIEW

**Review by dain — APPROVED** [review]
Looks good to me

**Comment by erichwang** [comment]
@electrum remove orphan files has the same pattern

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 1]
```diff
@@ -1123,7 +1122,10 @@ private void executeExpireSnapshots(ConnectorSession session, IcebergTableExecut
                 IcebergSessionProperties.EXPIRE_SNAPSHOTS_MIN_RETENTION);
 
         long expireTimestampMillis = session.getStart().toEpochMilli() - retention.toMillis();
-        expireSnapshots(table, expireTimestampMillis, session, executeHandle.getSchemaTableName());
+
+        table.expireSnapshots()
```
Comment (homar) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1126]: So actually I did it the same way originally but then I got this comment from @rdblue 
https://github.com/trinodb/trino/pull/10810#issuecomment-1044765060

**Comment by homar** [comment]
> @electrum remove orphan files has the same pattern

it is also based on the `remove_orphan_files` from iceberg's spark extension

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java (hunk 1 above)
Comment (homar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1126]: current implementation is based on iceberg's spark extention expire_snapshots

**Review by findepi — DISMISSED** [review]
Per https://github.com/trinodb/trino/pull/13399/files#r933007638 this is not the way to do this.

cc @alexjo2144

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 2]
```diff
@@ -1123,7 +1122,10 @@ private void executeExpireSnapshots(ConnectorSession session, IcebergTableExecut
...
+                .expireOlderThan(expireTimestampMillis)
+                .commit();
```
Comment (homar) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1128]: btw what you did is pretty much the same as what I called few lines below, see 
`table.expireSnapshots().expireOlderThan(expireTimestamp).cleanExpiredFiles(false).commit();` 
See line 1169.

**Comment by electrum** [comment]
@rdblue is using the Iceberg library in this manner ok? The implementation seems to cover cherry-picked commits, which the original code didn’t handle.

**Pull request merged**